packages/service-library/src/servicelib/aiohttp/long_running_tasks/client.py

```diff
@@ -0,0 +1,155 @@
+import asyncio
+from dataclasses import dataclass
+from typing import Any, AsyncGenerator, Coroutine, Final, Optional
+
+from aiohttp import ClientConnectionError, ClientSession, web
+from pydantic import Json
+from tenacity import TryAgain, retry
+from tenacity._asyncio import AsyncRetrying
+from tenacity.retry import retry_if_exception_type
+from tenacity.stop import stop_after_delay
+from tenacity.wait import wait_fixed, wait_random_exponential
+from yarl import URL
+
+from ..rest_responses import unwrap_envelope
+from .server import TaskGet, TaskId, TaskProgress, TaskStatus
+
+RequestBody = Json
+
+_MINUTE: Final[int] = 60
+_HOUR: Final[int] = 60 * _MINUTE
+
+
+_DEFAULT_AIOHTTP_RETRY_POLICY = dict(
+    retry=retry_if_exception_type(ClientConnectionError),
+    wait=wait_random_exponential(max=20),
+    stop=stop_after_delay(30),
+    reraise=True,
+)
+
+
+@retry(**_DEFAULT_AIOHTTP_RETRY_POLICY)
+async def _start(
+    session: ClientSession, request: URL, json: Optional[RequestBody]
+) -> TaskGet:
+    async with session.post(request, json=json) as response:
+        response.raise_for_status()
+        data, error = unwrap_envelope(await response.json())
+    assert not error  # nosec
+    assert data is not None  # nosec
+    task = TaskGet.parse_obj(data)
+    return task
+
+
+async def _wait_for_completion(
+    session: ClientSession,
+    task_id: TaskId,
+    status_url: str,
+    wait_timeout_s: int,
+    wait_interval_s: float,
+) -> AsyncGenerator[TaskProgress, None]:
+    try:
+        async for attempt in AsyncRetrying(
+            wait=wait_fixed(wait_interval_s),
+            stop=stop_after_delay(wait_timeout_s),
+            reraise=True,
+        ):
+            with attempt:
+                async with session.get(status_url) as response:
+                    response.raise_for_status()
+                    data, error = unwrap_envelope(await response.json())
+                    assert not error  # nosec
+                    assert data is not None  # nosec
+                task_status = TaskStatus.parse_obj(data)
+                yield task_status.task_progress
+                if not task_status.done:
+                    raise TryAgain(
+                        f"{task_id=}, {task_status.started=} has "
+                        f"status: '{task_status.task_progress.message}'"
+                        f" {task_status.task_progress.percent}%"
+                    )
+    except TryAgain as exc:
+        # this is a timeout
+        raise asyncio.TimeoutError(
+            f"Long running task {task_id}, calling to {status_url} timed-out after {wait_timeout_s} seconds"
+        ) from exc
+
+
+@retry(**_DEFAULT_AIOHTTP_RETRY_POLICY)
+async def _task_result(session: ClientSession, result_href: str) -> Any:
```

Comment:
question, why should I doc a private method? I will add the doc to the public one if that's ok with you.

Reply:
I **only doc what exceptions might raise** since it he only thing that to my knowledge programatically we cannot do ... 
Then when I use it , i aggregate them all nicely in the next-level function and so on ...